superset/tasks/alerts/oberver.py

```diff
@@ -0,0 +1,104 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import logging
+from datetime import datetime
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+
+from superset import db
+from superset.models.alerts import SQLObservation, SQLObserver
+from superset.sql_parse import ParsedQuery
+
+logger = logging.getLogger("tasks.email_reports")
+
+
+def observe(alert_id: int) -> Optional[str]:
+    """
+    Runs the SQL query in an alert's SQLObserver and then
+    stores the result in a SQLObservation.
+    Returns an error message if the observer value was not valid
+    """
+
+    sql_observer = db.session.query(SQLObserver).filter_by(alert_id=alert_id).one()
+
+    value = None
+    valid_result = True
+
+    parsed_query = ParsedQuery(sql_observer.sql)
+    sql = parsed_query.stripped()
+    df = sql_observer.database.get_df(sql)
+
+    error_msg = check_observer_result(df, sql_observer.id, sql_observer.name)
+
+    if error_msg:
+        valid_result = False
```

Comment:
let's have error_msg field in the SQLObservation instead of the valid result.
it will help user to understand why observation failed
error can be either from the database or from your custom validation